src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+        private void Write(ReadOnlySpan<char> value)
+        {
+            int written;
+            while (!_output.Encoder.TryEncode(value, _output.Buffer, out int consumed, out written))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteAscii(ReadOnlySpan<char> values)
         {
-            if (_prettyPrint) {
-                for (int i = 0; i < _indent; i++) {
-                    _formatter.Append("    ");
+            if (UseFastUtf8)
+            {
+                int len = values.Length;
+                Span<byte> buffer = _output.Buffer;
+                if (buffer.Length < len)
+                {
+                    _output.Enlarge(len);
+                    buffer = _output.Buffer;
                 }
+
+                ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
+                ref char chars = ref values.DangerousGetPinnableReference();
+                for (var i = 0; i < len; i++)
+                    Unsafe.Add(ref utf8Bytes, i) = (byte)Unsafe.Add(ref chars, i);
+            }
+            else if (UseFastUtf16)
+            {
+                int needed = values.Length * sizeof(char);
+                Span<byte> buffer = _output.Buffer;
+                if (buffer.Length < needed)
+                {
+                    _output.Enlarge(needed);
+                    buffer = _output.Buffer;
+                }
+
+                Span<char> span = buffer.NonPortableCast<byte, char>();
+                values.CopyTo(span);
             }
+            else
+            {
+                Write(values);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteItemSeperator()
+        {
+            if (_firstItem) return;
+
+            WriteChar(JsonConstants.ListSeperator);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteSpacing(bool newLine = true)
+        {
+            if (!_prettyPrint) return;
+
+            if (UseFastUtf8)
+                WriteSpacingUtf8(newLine);
+            else if (UseFastUtf16)
+                WriteSpacingUtf16(newLine);
+            else
+                WriteSpacingSlow(newLine);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteSpacingSlow(bool newLine)
+        {
+            if (newLine)
+            {
+                WriteChar(JsonConstants.CarriageReturn);
+                WriteChar(JsonConstants.LineFeed);
+            }
+
+            int indent = _indent;
+            while (indent-- >= 0)
+            {
+                WriteChar(JsonConstants.Space);
+                WriteChar(JsonConstants.Space);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteSpacingUtf8(bool newline)
```

Comment:
This seems like a very long method to force its inlining. Also, is it even that hot?